### PR TITLE
feat(schema): seeded bootstrap, Mann-Whitney U and Kolmogorov-Smirnov (ENG-146)

### PR DESCRIPTION
**Pull request by dbworku**
**ENG-146 — rank the leaderboard inferentially, not on a bare median.**
Split into a 6-PR stack (merge bottom-up, each branch independently green on its own):

1. #116 — ci: dispatch `bench-matrix` per-provider + fix two runner-cache/publish bugs
2. **#114 — schema:** seeded bootstrap, Mann-Whitney U, Kolmogorov-Smirnov (pure-additive toolkit) ← _you are here_
3. #115 — results: share a rank on statistical ties; create the `LEADERBOARD.md` drift gate
4. #117 — repo-checks: render the leaderboard from the committed dataset, not the raw tree
5. #118 — results: render the `p (KS)` column; stop the gate silencing its own guard
6. #119 — dataset: publish run `29061549181` + the regenerated leaderboard

↑ **This PR is #114 (2/6)**, based on #116. Nothing here consumes the toolkit yet — the wiring lands in #115.

---

## Why the leaderboard needs statistics at all

A provider's `Samples` are repeated trials **inside one sandbox**. Their spread is therefore *environmental* — noisy neighbours, host contention, virtualization overhead — not measurement error that shrinks as we collect more trials. Two consequences follow, and both are load-bearing:

- The spread does **not** vanish with more trials, so a bare [median](https://en.wikipedia.org/wiki/Median) is a point estimate carrying real, unreported uncertainty.
- Ranking on that bare median lets a lucky draw buy a position. In a live run modal's STREAM `Copy` spanned **9.7k–65k MB/s** while daytona sat at **66.5k ±0.14%**. Sorting on the median alone can put either one first.

This PR adds the tools to say *how confident* a ranking is. Nothing consumes them yet — the wiring lands in #115 — so each is unit-tested directly.

### `bootstrapMedianInterval` — a [confidence interval](https://en.wikipedia.org/wiki/Confidence_interval) around the median

A [percentile bootstrap](https://en.wikipedia.org/wiki/Bootstrapping_(statistics)#Methods_for_bootstrap_confidence_intervals): resample the trials with replacement 10,000 times, take each resample's median, and report the empirical 2.5th/97.5th percentiles of that distribution.

- **The median, not the mean** — it is [robust](https://en.wikipedia.org/wiki/Robust_statistics): one stalled pass drags a mean far more than it moves a median, and we want the provider's *typical* throughput, not its average-including-the-stall.
- **The [bootstrap](https://en.wikipedia.org/wiki/Bootstrapping_(statistics)), not a normal-theory `±1.96·σ/√n`** — that closed form assumes a symmetry the data plainly lacks (see modal's 9.7k…65k range) and independence from the host's scheduler, which is exactly what these samples violate.
- At `n = 1` it degenerates to a point interval rather than fabricating a bound by resampling one value.

### `mannWhitneyU` — are two providers actually different?

The [Mann–Whitney U test](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test) (two-sided, normal approximation with tie and [continuity](https://en.wikipedia.org/wiki/Continuity_correction) corrections) asks precisely the leaderboard's question: do these two samples come from one distribution, or does one genuinely tend to be faster?

It is [non-parametric](https://en.wikipedia.org/wiki/Nonparametric_statistics) and rank-based, so it assumes neither normality nor equal variance — both violated here — and a single catastrophic pass moves a rank by one position instead of dragging a mean.

⚠️ At this suite's `n ≈ 5` the smallest attainable two-sided [p-value](https://en.wikipedia.org/wiki/P-value) is ≈ 0.008. A real difference is detectable, but only a large one. **A non-significant result means "not enough evidence", never "the providers are equal"** — the classic [type II error](https://en.wikipedia.org/wiki/Type_I_and_type_II_errors).

### `kolmogorovSmirnov` — do the *shapes* differ?

The [two-sample Kolmogorov–Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test) reports the largest gap between the two [empirical CDFs](https://en.wikipedia.org/wiki/Empirical_distribution_function). Where U tests a shift in central tendency, D also catches a provider whose median coincides with another's but whose distribution is [bimodal](https://en.wikipedia.org/wiki/Multimodal_distribution) — alternating between fast and stalled passes. **That bimodality is exactly what environmental noise looks like**, so it is reported alongside U rather than driving the ranking.

`kolmogorovQ` sums the asymptotic tail series *to convergence*. The series alternates (`1 − 1 + 1 − 1 …`), so a fixed-term truncation collapses toward 0 at small λ and would report p ≈ 0 — "certainly different" — for two **identical** samples.

### `seededRng` — because the leaderboard is a committed artifact

`LEADERBOARD.md` is checked into the repo. A `Math.random()`-driven bootstrap would rewrite every confidence bound on each regeneration, making the diff meaningless and the artifact non-[reproducible](https://en.wikipedia.org/wiki/Reproducibility).

So the [PRNG](https://en.wikipedia.org/wiki/Pseudorandom_number_generator) (mulberry32) is seeded by hashing stable Run/Metric/provider identity with [FNV-1a](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function). A given Run renders byte-identically on every machine and every rerun — which is also what makes #115's drift gate sound rather than flaky.

## Correctness

Two defects found in review, both verified against the code rather than taken on report:

| | Symptom | Cause |
|---|---|---|
| **P1** | `kolmogorovSmirnov([NaN, 1], [1, 2])` **never returns** (had to be `SIGKILL`ed) | the two-pointer ECDF sweep advances on `x <= y` / `y <= x`; both are `false` against `NaN`, so neither index moves |
| **P1** | `mannWhitneyU([NaN, 1, 2], [3, 4, 5])` returned a plausible `p ≈ 0.081` | `NaN` has no order, silently corrupting the sort-based midranks |

Both consume samples *by order*, and `NaN` has no order. Rather than adding a fourth copy of the guard, the one `aggregate`/`percentileOf`/`bootstrapMedianInterval` each already spelled out was extracted into `assertFiniteSamples`. `options.resamples` is guarded too: `0`/`NaN` previously produced `lo`/`hi` of `NaN`, which serialize into the committed leaderboard as `null` instead of failing.

The KS test added here **hangs without the fix** — a real regression guard, not a style check.

---
## Summary by cubic
Adds seeded bootstrap and two nonparametric tests to quantify sample noise and avoid lucky median rankings. Preps leaderboard inference for ENG-146; pure additive utilities, not wired yet.

- **New Features**
  - `seededRng`: deterministic mulberry32 seeded by an FNV-1a (UTF-16) variant to keep the committed leaderboard stable.
  - `bootstrapMedianInterval`: percentile bootstrap of the median (10k resamples; supports `level`, `seed`; point interval at n=1).
  - `mannWhitneyU` and `kolmogorovSmirnov`: two-sample tests with tie/continuity corrections; KS tail converges and is anti-conservative at tiny n; return statistic and two-sided p-value.
  - `DEFAULT_ALPHA = 0.05`: exported default two-sided significance level for tests.

- **Bug Fixes**
  - Reject non-finite samples, empty inputs, and invalid `resamples`; options validated even at n=1.
  - Prevent KS hangs on `NaN` and MWU from returning misleading p-values; identical samples correctly yield p = 1 via a convergent KS series.

<sup>Written for commit cf186a040578dda10d52d3eb634e4fcd69c27ac0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/starslingdev/sandbox-benchmarks/pull/114?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added statistical inference utilities for comparing repeated measurements, including deterministic seeded bootstrapped median confidence intervals.
  * Added two-sample distribution tests (Mann–Whitney U and Kolmogorov–Smirnov) with configurable confidence level and default two-sided significance.
* **Bug Fixes**
  * Improved robustness by validating inputs and rejecting empty or non-finite samples to avoid incorrect results and potential hangs.
* **Tests**
  * Expanded coverage for seeded determinism, interval edge cases, tie/identical-sample handling, symmetry, and p-value/behavior expectations.

### CODE REVIEW

**Comment by dbworku** [comment]
* **#131** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/131?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#130** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/130?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#129** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/129?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#123** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/123?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#121** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#119** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#118** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/117?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#115** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/115?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#114** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/114?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/114?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#116** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/116?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>merge-ready</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/starslingdev?ref=merge-queue-instructions-in-stack-comment&prId=114).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by dbworku** [comment]
Both cubic findings were real and are fixed in `fix(schema): reject non-finite Samples and bad resample counts`. I verified each against the code rather than taking the report at face value:

**P1 — `kolmogorovSmirnov` hung on `NaN`.** Confirmed empirically: `kolmogorovSmirnov([NaN, 1], [1, 2])` never returns and had to be `SIGKILL`ed. The two-pointer ECDF sweep advances on `x <= y` / `y <= x`, and both are `false` against `NaN`, so neither index moves. A single bad Sample would have hung the leaderboard build outright.

**P1 — `mannWhitneyU` accepted `NaN` silently.** `mannWhitneyU([NaN, 1, 2], [3, 4, 5])` returned `p ≈ 0.081`. That's arguably worse than the hang: the number looks real and would have driven a published ranking.

**P2 — `resamples` unvalidated.** Slightly worse than reported: `resamples: 0` and `NaN` return `lo`/`hi` of `NaN`, which serialize into the *committed* `LEADERBOARD.md` as `null` rather than failing. `-1` threw a bare `RangeError` naming neither the caller nor the value.

Rather than adding a fourth copy of the guard, I extracted the one `aggregate`, `percentileOf` and `bootstrapMedianInterval` already each spelled out into `assertFiniteSamples`, and applied it to both two-sample tests. The `n = 1` point-interval path returns before the `resamples` guard and is unaffected.

The KS test added here **hangs without the fix** — it's a real regression guard, not a style check.

**Comment by dbworku** [comment]
Both findings confirmed and fixed.

**`resamples` validated after the `n=1` early return** — you're right, and it was inconsistent with `level`. Reproduced:

```
n=1, resamples:0   -> OK  {"median":42,...,"resamples":0}
n=1, resamples:-1  -> OK
n=1, level:0       -> THROWS
n=2, resamples:0   -> THROWS
```

All options are now checked before any early return, so a misconfigured call is rejected regardless of how many Samples it carries. The `n=1` point interval still returns, just once the options are known good. Test added for exactly this.

**FNV-1a code units vs bytes** — correct, and I took the docstring option deliberately rather than the re-encode. `charCodeAt` yields 16-bit units, so this matches textbook FNV-1a only for ASCII seeds, which is all it is ever given (`<runId>:<metricId>:<providerId>`). Nothing needs to interoperate with another FNV-1a implementation; the requirement is only a stable, well-mixed 32-bit seed. Encoding to UTF-8 bytes would change the hash and therefore **silently rewrite every confidence bound in the committed `LEADERBOARD.md`** — churn with no benefit. The comment now says what the code actually does, and warns that it diverges above U+007F.

Verified `seededRng`'s draws are bit-identical after the change, and #115's drift gate still passes, so the committed artifact is untouched.

File: packages/schema/src/analysis.ts [hunk 1]
```diff
@@ -104,3 +104,305 @@ export function aggregate(samples: number[]): Aggregates {
...
+		const y = sortedB[j] as number;
+		// Advance past every copy of the smaller value (both, when equal) before measuring the gap, so
+		// ties can't report a spurious step.
+		if (x <= y) while (i < nA && sortedA[i] === x) i++;
+		if (y <= x) while (j < nB && sortedB[j] === y) j++;
+		statistic = Math.max(statistic, Math.abs(i / nA - j / nB));
+	}
+
+	const en = Math.sqrt((nA * nB) / (nA + nB));
+	return { statistic, pValue: kolmogorovQ((en + 0.12 + 0.11 / en) * statistic), nA, nB };
```
Comment (cosmic-emerald) [inline packages/schema/src/analysis.ts:384]: for small sample sizes, e.g. `[1,2,3]` vs `[4,5,6]`it returns ≈0.03262, while the exact two-sided p-value is `2 / C(6,3) = 0.1`

**Review by cosmic-emerald — APPROVED** [review]

File: packages/schema/src/analysis.ts (hunk 1 above)
Comment (dbworku) [reply packages/schema/src/analysis.ts:384]: Will have Fable double check the math. I was definitely trusting it on stats here.

Comment (dbworku) [reply packages/schema/src/analysis.ts:384]: Documented.
